codaco/Server
# Lay groundwork for device pairing

This is a prototype of the new server requirements & takes care of a lot of the initial work for #41. Includes:

- http/rest-based API for devices
- http/rest-based API for admin/gui
- pairing modals are hooked up to API + redux states

Browser + bare server dev is still supported, but there's a new recommended electron workflow in the readme (which I'll gradually improve).

### Testing device pairing:

1. Run the dev app: `npm start:electron-renderer` and `npm run start:electron`
1. Open the main dashboard UI
1. Initiate a pairing request, e.g.: `curl localhost:51001/devices/new`
    - expected: API response containing request ID.
    - expected: in GUI, see pairing prompt (with dismiss & pair buttons)
1. Click Pair with Device
    - expected: see modal with alphanumeric pairing code
1. Validate the pairing request (as though you entered code on a client device)
    - e.g.: `curl -H "Content-Type: application/json" -d '{"requestId":"01d30d6c-e134-443f-b488-431d5b5360a4", "pairingCode":"3kkEDBS0cVhQ"}' localhost:51001/devices/`
    - Note you can select & copy the pairing code from the GUI display boxes
    - expected: API success response
    - expected: GUI confirmation ("all set" modal)
    - If request ID or code were invalid, expect 4xx API response and no GUI change.

### Known issues

- API / model objects will likely change
- UI 'finish' (e.g., modal transitions)
- no crypto on key transmission yet

... and more... as this is a fairly significant departure based on the server changes doc, I'd like to get feedback on the general direction rather than wait until a feature is complete.
